shared/modPersistentData.js

```diff
@@ -0,0 +1,1047 @@
+	{
+		await asyncDBRun( "UPDATE accounts SET currentAuthToken = ?, currentAuthTokenExpirationTime = ? WHERE id = ?", [ token, Date.now() + TOKEN_EXPIRATION_TIME, id ] )
+	},
+
+	AsyncUpdatePlayerUsername: async function AsyncUpdatePlayerUsername( id, username )
+	{
+		await asyncDBRun( "UPDATE accounts SET username = ? WHERE id = ?", [ username, id ] )
+	},
+
+	AsyncUpdatePlayerAuthIp: async function AsyncUpdatePlayerAuthIp( id, lastAuthIp )
+	{
+		await asyncDBRun( "UPDATE accounts SET lastAuthIp = ? WHERE id = ?", [ lastAuthIp, id ] )
+	},
+
+	AsyncUpdatePlayerCurrentServer: async function AsyncUpdatePlayerCurrentServer( id, serverId )
+	{
+		await asyncDBRun( "UPDATE accounts SET currentServerId = ? WHERE id = ?", [ serverId, id ] )
+	},
+
+	AsyncWritePlayerPersistenceBaseline: async function AsyncWritePlayerPersistenceBaseline( id, persistentDataBaseline )
+	{
+		await asyncDBRun( "UPDATE accounts SET persistentDataBaseline = ? WHERE id = ?", [ persistentDataBaseline, id ] )
+	},
+
+	AsyncGetPlayerModPersistence: async function AsyncGetPlayerModPersistence( id, pdiffHash )
+	{
+		// prevent JSON parse problems when the user has no data in the database
+		let result = await asyncDBGet( "SELECT data from modPersistentData WHERE id = ? AND pdiffHash = ?", [ id, pdiffHash ] )
+		if ( result == undefined )
+		{
+			await asyncDBRun( "INSERT INTO modPersistentData ( id, pdiffHash, data ) VALUES ( ?, ?, ? )", [ id, pdiffHash, "{}" ] )
+			result = await asyncDBGet( "SELECT data from modPersistentData WHERE id = ? AND pdiffHash = ?", [ id, pdiffHash ] )
+		}
+		return JSON.parse( result.data )
+	},
+
+	AsyncWritePlayerModPersistence: async function AsyncWritePlayerModPersistence( id, pdiffHash, data )
+	{
+		await asyncDBRun( "UPDATE modPersistentData SET data = ? WHERE id = ? AND pdiffHash = ?", [ data, id, pdiffHash ] )
+		console.log( "successfully written pdiff data" )
+	},
+
+	AsyncModPersistenceBufferToJson: async function AsyncModPersistenceBufferToJson( modInfo, playerID, buffer )
+	{
+		// this returns an object in the form
+		/*
+		{
+			baseline: <baseline> // this is the vanilla persistence, that we will write like normal (as a buffer)
+			pdiffs: // array of all the mods and the persistence data they have stored
+			[
+				{
+					hash: <hash>, // hashed string 
+					data: <data>,  // Object
+					pdef: <pdef> // Object (just used for temp storage tbh)
+				},
+				{
+					hash: <hash>, // hashed string 
+					data: <data>,  // Object
+					pdef: <pdef> // Object (just used for temp storage tbh)
+				},
+				...
+			]
+		}
+		*/
+
+		let ret = {
+			pdiffs: []
+		}
+
+		let pdiffs = modInfo.Mods.filter( m => !!m.Pdiff ).map( m => m.Pdiff )
+
+		var pdefCopy = JSON.parse( JSON.stringify( DEFAULT_PDEF_OBJECT ) )
+
+		for ( let pdiffstr of pdiffs )
+		{
+			let pdiff
+			if ( pdiffstr )
+			{
+				try
+				{
+					let pdiffHash = crypto.createHash( "sha1" ).update( pdiffstr ).digest( "hex" )
```

Comment:
Thinking of changing this to hash the mod name, because if a mod updates and adds anything to their pdiff file, all of the data for the users of the mod will be lost and waste space in the DB unnecessarily